yjit/src/invariants.rs

```diff
@@ -15,10 +15,25 @@ use std::collections::HashMap;
 // assume_single_ractor_mode(jit)
 // assume_stable_global_constant_state(jit);
 
+struct MethodLookupDependency {
+    block: BlockRef,
+    mid: ID
+}
+
 /// Used to track all of the various block references that contain assumptions
 /// about the state of the virtual machine.
 pub struct Invariants {
-    bops: HashMap<(RedefinitionFlag, ruby_basic_operators), Vec<BlockRef>>
+    /// Tracks block assumptions about the stability of a basic operator on a
+    /// given class.
+    basic_operators: HashMap<(RedefinitionFlag, ruby_basic_operators), Vec<BlockRef>>,
+
+    /// Tracks block assumptions about callable method entry validity.
+    cme_validity: HashMap<*const rb_callable_method_entry_t, Vec<BlockRef>>,
+
+    /// Tracks block assumptions about method lookup. Maps a class to a table of
+    /// method ID points to a set of blocks. While a block `b` is in the table,
+    /// b->callee_cme == rb_callable_method_entry(klass, mid).
+    method_lookup: HashMap<VALUE, HashMap<ID, Vec<MethodLookupDependency>>>
```

Comment:
We might want to use class serials here for the key? Classes are movable by the Ruby GC if I'm not mistaken? @XrXr

Reply:
The intent is certainly to be able to move them long-term, even if they aren't now. We'd love for compaction to work on everything...